File: src-tauri/src/app/helpers/factories/manifest_factory.rs
use crate::app::helpers::factories::ModFactory;
use crate::app::Config;
use indexmap::IndexMap;
use serde::Serialize;
use std::fs;

#[cfg(test)]
pub struct ManifestFactory {
    mc_version: String,
    mods: Vec<ModFactory>,
}

#[cfg(test)]
impl ManifestFactory {
    pub fn new(mc_version: impl Into<String>) -> Self {
        Self {
            mc_version: mc_version.into(),
            mods: Vec::new(),
        }
    }

    pub fn with_mod(mut self, m: &ModFactory) -> Self {
        self.mods.push(m.clone());
        self
    }

    pub fn with_mods(mut self, mods: &Vec<ModFactory>) -> Self {
        self.mods = mods.clone();
        self
    }

    pub fn write(self) {
        let manifest_mods: IndexMap<String, String> = self
            .mods
            .iter()
            .map(|m| (m.id.clone(), m.version.clone()))
            .collect();

        let manifest = TestManifest {
            name: "Pack",
            version: "1.0.0",
            side: "both",
            modloader: "fabric",
            minecraft_version: &self.mc_version,
            default_provider: "modrinth",
            mods: &manifest_mods,
        };

        fs::write(
            Config::manifest_path(),
            serde_json::to_string_pretty(&manifest).unwrap(),
        )
            .unwrap();
    }
}

#[cfg(test)]
#[derive(Serialize)]
struct TestManifest<'a> {
    name: &'a str,
    version: &'a str,
    side: &'a str,
    modloader: &'a str,
    minecraft_version: &'a str,
    default_provider: &'a str,
    mods: &'a IndexMap<String, String>,
}

File: packages/core/src/testing/manifest-factory.ts
import {writeFileSync} from "fs";
import type {ConfigPaths} from "../models/config.js";
import type {ResourceKind} from "../models/manifest.js";
import type {ModFactory} from "./mod-factory.js";

/** Factory for creating test mcpm.json files */
export class ManifestFactory {
  private mcVersion: string;
  private byKind: Record<ResourceKind, ModFactory[]> = { mod: [], datapack: [], resourcepack: [], shaderpack: [] };

  private constructor(mcVersion: string) {
    this.mcVersion = mcVersion;
  }

  static create(mcVersion: string): ManifestFactory {
    return new ManifestFactory(mcVersion);
  }

  withMod(m: ModFactory): this {
    this.byKind.mod.push(m);
    return this;
  }

  withMods(mods: ModFactory[]): this {
    this.byKind.mod = [...mods];
    return this;
  }

  withDatapack(d: ModFactory): this {
    this.byKind.datapack.push(d);
    return this;
  }

  withDatapacks(datapacks: ModFactory[]): this {
    this.byKind.datapack = [...datapacks];
    return this;
  }

  withResourcepack(r: ModFactory): this {
    this.byKind.resourcepack.push(r);
    return this;
  }

  withResourcepacks(resourcepacks: ModFactory[]): this {
    this.byKind.resourcepack = [...resourcepacks];
    return this;
  }

  withShaderpack(s: ModFactory): this {
    this.byKind.shaderpack.push(s);
    return this;
  }

  withShaderpacks(shaderpacks: ModFactory[]): this {
    this.byKind.shaderpack = [...shaderpacks];
    return this;
  }

  writeTo(paths: ConfigPaths): void {
    const manifest = {
      name: "Pack",
      version: "1.0.0",
      side: "both",
      modloader: "fabric",
      minecraft_version: this.mcVersion,
      default_provider: "modrinth",
      mods: toVersionObj(this.byKind.mod),
      datapacks: toVersionObj(this.byKind.datapack),
      resourcepacks: toVersionObj(this.byKind.resourcepack),
      shaderpacks: toVersionObj(this.byKind.shaderpack),
    };
    writeFileSync(paths.manifestPath, JSON.stringify(manifest, null, 2));
  }
}

function toVersionObj(items: ModFactory[]): Record<string, string> {
  const obj: Record<string, string> = {};
  for (const m of items) {
    obj[m.id] = m.version;
  }
  return obj;
}
